---
title: "P89 RNA-seq CAR transcript analysis"
author: "James Eddy"
date: "March 30, 2016"
output: html_document
---

```{r global_opts, echo=FALSE}
knitr::opts_chunk$set(fig.width=8.5, fig.align='center',
                      echo=FALSE, warning=FALSE, message=FALSE)
# knitr::opts_knit$set(root.dir = "..")
```

```{r load_packages}
library(stringr)
library(tidyr)
library(dplyr)
library(ggplot2)
library(viridis)
library(rtracklayer)
library(ggthemes)
library(scales)
library(cowplot)

myCbPal <- colorblind_pal()(8)
myCbPal[c(1, 3, 6)] <- myCbPal[c(6, 1, 3)]
myCbPal[3] <- "#666666"
```

```{r load_data}
load("data/sample_metrics_data.RData")
load("data/sample_rapmap_data.RData")
load("data/sample_salmon_data.RData")
load("data/sample_tcr_data.RData")

gff_file <- "~/Box Sync/data/projects/carTcellAnalysis/annotation/carPlus.gtf"
xcripts_gtf <- import.gff2(gff_file)
```

```{r prep_gtf_dat}
xcript_dat <- as.data.frame(xcripts_gtf)

car_dat <- xcript_dat %>% 
    filter(seqnames == "CAR-1") %>% 
    mutate(segment = factor(transcript_id, levels = transcript_id))
```

```{r format_cov_dat}
# function to append CAR segment
get_segment <- Vectorize(function(pos) {
    xcript_dat %>% 
        filter(seqnames == "CAR-1",
               start < pos & end > pos) %>% 
        select(transcript_id) %>% 
        as.character()
})

lib_cov <- cov_dat %>% 
    filter(lib_id == cov_dat$lib_id[1]) %>% 
    mutate(segment = get_segment(pos)) %>% 
    select(pos, segment)

cov_dat <- cov_dat %>% 
    left_join(lib_cov, by = c("pos" = "pos"))
```


```{r data_clean_fxns}
# clean up duplicated headers
clean_dup_names <- function(df) {
    df_names <- names(df)
    is_dup_name <- duplicated(df_names)
    df_names[is_dup_name] <- str_c(df_names[is_dup_name], "2")
    names(df) <- df_names
    return(df)
}

# simplify library ID
simplify_lib_id <- function(df) {
    df_names <- names(df)
    is_lib_name <- str_detect(tolower(df_names), "lib.*id")
    df_names[is_lib_name] <- "lib_id"
    names(df) <- df_names
    
    df %>% 
        mutate(lib_id = str_extract(lib_id, "lib[0-9]+"))
}

# clean/relabel donor ID
clean_donor_ids <- function(df) {
    df %>% 
        mutate(donor_id = tolower(donor_id))
}

# relabel timepoints
relabel_timepoints <- function(df) {
    df %>% 
        mutate(timepoint = str_replace(timepoint, " ", ""),
               timepoint = str_replace(timepoint, "InfusionProduct", "IP"),
               timepoint = str_replace(timepoint, "Day0", "t0"),
               timepoint = str_replace(timepoint, "Day(7|8|9|12)", "t1"),
               timepoint = str_replace(timepoint, "Day(26|28|29|33)", "t2"))
}

relabel_transcripts <- function(df, xcript_dat) {
    df_names <- names(df)
    for (i in 1:length(df_names)) {
        if (df_names[i] %in% xcript_dat$seqnames) {
            xcript_row <- which(xcript_dat$seqnames %in% df_names[i])
            new_name <- xcript_dat$transcript_id[xcript_row]
            df_names[i] <- new_name
        }
    }
    names(df) <- df_names
    return(df)
}
```

```{r data_prep_fxn}
prep_cov_dat <- function(lib_dat, metric_dat) {
    lib_dat %>% 
        clean_dup_names() %>% 
        simplify_lib_id() %>%
        clean_donor_ids() %>% 
        select(lib_id, donor_id, timepoint) %>% 
        relabel_timepoints() %>% 
        left_join(cov_dat, by = c("lib_id" = "lib_id")) %>% 
        left_join(metric_dat %>% 
                      clean_dup_names() %>% 
                      simplify_lib_id() %>% 
                      select(lib_id, fastq_total_reads, 
                             median_cv_coverage, mapped_reads_w_dups),
                  by = c("lib_id" = "lib_id")) %>% 
        left_join(map_rate_dat, by = c("lib_id" = "lib_id")) %>% 
        mutate(mapped_reads = fastq_total_reads * map_rate,
               norm_cov = cov / mapped_reads) %>% 
        left_join(salmon_quant_dat %>% 
                      # filter(Name == "CAR-1") %>% 
                      select(lib_id, Name, TPM) %>% 
                      spread(Name, TPM) %>% 
                      dplyr::rename(CAR = `CAR-1`) %>% 
                      relabel_transcripts(xcript_dat)) %>% 
        left_join(jxn_summary_dat) %>% 
        mutate_each_(funs(. = ifelse(is.na(.), 0, .)), 
                     vars = c("tra_pos", "trb_pos", "tcr_pos"))    
}
```


```{r plot_fxn}
plot_coverage <- function(formatted_cov_dat, gtf_dat, 
                          split = TRUE, fits_only = FALSE, 
                          color_by = "lib_num", fade_by = "log2(CAR + 1)",
                          hide_legend = FALSE) {
    
    # add lib numbers for coloring
    formatted_cov_dat <- formatted_cov_dat %>% 
        group_by(donor_id, timepoint) %>% 
        mutate(lib_num = dense_rank(lib_id) / n_distinct(lib_id),
               lib_num = as.character(lib_num)) %>% 
        ungroup()

    # determine plot height
    height <- log2(max(formatted_cov_dat$cov, na.rm = TRUE) + 1)
    
    # create color scale
    if (!fits_only) {
        gradient_stop <- myCbPal[3]
    } else {
        gradient_stop <- myCbPal[2]
    }
    num_libs <- n_distinct(formatted_cov_dat$lib_id)
    cc <- seq_gradient_pal(myCbPal[3], gradient_stop)(seq(0, 1, length.out = num_libs))
    
    # build plot
    p_cov <- ggplot() +
        geom_rect(data = gtf_dat, 
                  aes(xmin = start, xmax = end, ymin = 0, ymax = height, 
                      fill = segment),
                  alpha = 0.5, colour = "gray")
    if (!fits_only) {
        p_cov <- p_cov +
            geom_point(data = formatted_cov_dat, 
                       aes_string(x = "pos", y = "log2(cov + 1)", 
                                  alpha = fade_by, colour = color_by),
                       stroke = 0) +
            geom_smooth(data = formatted_cov_dat,
                        aes(x = pos, y = log2(cov + 1)),
                        se = FALSE, colour = myCbPal[2])
    } else {
        p_cov <- p_cov +
            geom_line(data = formatted_cov_dat,
                      aes_string(x = "pos", y = "log2(cov + 1)", 
                                 group = "lib_id", colour = color_by,
                                 alpha = fade_by),
                      stat = "smooth", method = "loess", 
                      se = FALSE, size = 1)
    }
    
    if (color_by == "lib_num") {
        p_cov <- p_cov + 
            scale_color_manual(values = cc) +
            guides(colour = FALSE)        
    } else {
        p_cov <- p_cov +
            scale_color_gradient(low = myCbPal[3], high = myCbPal[2])
    }
    
    if (is.numeric(fade_by)) {
        p_cov <- p_cov +
            scale_alpha_continuous(range = c(fade_by, fade_by)) +
            guides(alpha = FALSE)
    } else {
        p_cov <- p_cov +
            scale_alpha_continuous(range = c(0.2, 0.8))
    }
    
    if (color_by == fade_by) {
        p_cov <- p_cov +
            guides(alpha = FALSE)
    }

    p_cov <- p_cov +    
        scale_fill_viridis(discrete = TRUE) +
        theme_gray()
    
    if (split) {
        p_cov <- p_cov +
            facet_grid(donor_id ~ timepoint)
    }
    
    if (hide_legend) {
        p_cov <- p_cov +
            guides(fill = FALSE)
    }
    return(p_cov)
}
```


```{r prep_bulk_cov_dat}
bulk_cov_dat <- bulk_lib_dat %>% 
    prep_cov_dat(bulk_metric_dat) %>% 
    filter(donor_id %in% c("x145", "x194", "x228"))
```

```{r plot_bulk_cov, fig.height=6}
plot_coverage(bulk_cov_dat, car_dat)
```

```{r prep_p89_c1_cov_dat}
p89_c1_cov_dat <- sc_lib_dat %>% 
    prep_cov_dat(sc_metric_dat)
```

```{r plot_p89_c1_cov, fig.height=6}
plot_coverage(p89_c1_cov_dat, car_dat)
```

```{r prep_p85_c1_cov_dat}
p85_c1_cov_dat <- p85_lib_dat %>% 
    prep_cov_dat(p85_metric_dat)
```

```{r plot_p85_c1_cov, fig.height=4}
plot_coverage(p85_c1_cov_dat, car_dat)
```

```{r p89_cov_fits, fig.height=6}
p89_c1_cov_dat %>% 
    plot_coverage(car_dat, fits_only = TRUE)
```

```{r p89_cov_exploration, fig.height=6}
p89_c1_cov_dat %>% 
    plot_coverage(car_dat, fits_only = TRUE, split = FALSE, 
                  color_by = "median_cv_coverage", fade_by = 0.7)
```

```{r p89_cov_jxn, fig.height=6}
p89_c1_cov_dat %>% 
    mutate(has_jxn = ifelse(tra_pos | trb_pos, 1, 0)) %>%
    plot_coverage(car_dat, fits_only = TRUE, split = FALSE, 
                  color_by = "has_jxn", fade_by = 0.7)
```

```{r p89_cov_jxn_only, fig.height=6}
p89_c1_cov_dat %>% 
    mutate(has_jxn = ifelse(tra_pos | trb_pos, 1, 0)) %>%
    filter(has_jxn > 0) %>% 
    plot_coverage(car_dat, fits_only = TRUE, split = FALSE)
```

```{r p89_cov_tcr, fig.height=6}
p89_c1_cov_dat %>% 
    mutate(has_tcr = ifelse(tcr_pos, 1, 0)) %>%
    plot_coverage(car_dat, fits_only = TRUE, split = FALSE, 
                  color_by = "has_tcr", fade_by = 0.7)
```

```{r p89_cov_tcr_only, fig.height=6}
p89_c1_cov_dat %>% 
    mutate(has_tcr = ifelse(tcr_pos, 1, 0)) %>%
    filter(has_tcr > 0) %>% 
    plot_coverage(car_dat, fits_only = TRUE, split = FALSE)
```

```{r p89_cov_car}
segment <- "CAR"
p89_c1_cov_dat %>% 
    plot_coverage(car_dat, fits_only = TRUE, split = FALSE,
                  hide_legend = TRUE, 
                  color_by = sprintf("log2(%s + 1)", segment),
                  fade_by = sprintf("log2(%s + 1)", segment))
```

```{r p89_all}
gene_list <-c("CAR", 
      xcript_dat$transcript_id %>% intersect(names(p89_c1_cov_dat))) %>% 
    as.list()
p_list <- lapply(gene_list, function(x) {
    p <- p89_c1_cov_dat %>% 
        plot_coverage(car_dat, fits_only = TRUE, split = FALSE, 
                      hide_legend = TRUE, 
                      color_by = sprintf("log2(%s + 1)", x),
                      fade_by = sprintf("log2(%s + 1)", x))
    return(p)
})
plot_grid(plotlist = p_list, ncol = 3, nrow = 5)
```

```{r}
indicator_xcripts <- c("CAR", "IgG4hinge_r1", "EGFRt_r2", "EGFRt_r4")
p89_c1_cov_dat %>% 
    select(one_of(indicator_xcripts)) %>% 
    rowwise() %>% 
    transmute(car_pos = as.numeric(any(log2(. + 1) > 5))) %>% 
    head() %>% select(car_pos)

%>% 
    bind_cols(p89_c1_cov_dat, .) %>% 
    plot_coverage(car_dat, fits_only = TRUE, split = FALSE, 
                  color_by = "car_pos", fade_by = 0.7)
```


```{r boxplot}
p89_c1_cov_dat %>% 
    mutate(segment = factor(segment, 
                            levels = xcript_dat$transcript_id)) %>% 
    filter(!is.na(segment)) %>% 
    ggplot(aes(x = segment)) +
    geom_bar(aes(colour = donor_id), position = "dodge")
```

